src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -5616,43 +5616,73 @@ private BoundExpression BindInitializerMemberAssignment(
+                            Debug.Assert((object)boundLeft.Type != null);
+
+                            var rhsExpr = initializer.Right.CheckAndUnwrapRefExpression(diagnostics, out RefKind refKind);
+                            bool isRef = refKind == RefKind.Ref;
+                            var rhsKind = isRef ? GetRequiredRHSValueKindForRefAssignment(boundLeft) : BindValueKind.RValue;
+
+                            // Bind member initializer value, i.e. right part of assignment
+                            BoundExpression boundRight = BindInitializerExpressionOrValue(
+                                syntax: rhsExpr,
+                                type: boundLeft.Type,
+                                rhsKind,
+                                typeSyntax: boundLeft.Syntax,
+                                diagnostics: diagnostics);
+
+                            // Bind member initializer assignment expression
+                            return BindAssignment(initializer, boundLeft, boundRight, isRef, diagnostics);
+                        }
+                        break;
+                    }
 
-                    var rhsExpr = initializer.Right.CheckAndUnwrapRefExpression(diagnostics, out RefKind refKind);
-                    bool isRef = refKind == RefKind.Ref;
-                    var rhsKind = isRef ? GetRequiredRHSValueKindForRefAssignment(boundLeft) : BindValueKind.RValue;
+                // We fall back on simply binding the name as an expression for proper recovery
+                // and also report a diagnostic about a simple identifier being an invalid expression
+                // in the object initializer, to indicate to the user that they are missing an assignment
+                case SyntaxKind.IdentifierName:
+                    {
+                        Error(diagnostics, ErrorCode.ERR_InvalidInitializerElementInitializer, memberInitializer);
 
-                    // Bind member initializer value, i.e. right part of assignment
-                    BoundExpression boundRight = BindInitializerExpressionOrValue(
-                        syntax: rhsExpr,
-                        type: boundLeft.Type,
-                        rhsKind,
-                        typeSyntax: boundLeft.Syntax,
-                        diagnostics: diagnostics);
+                        var identifierName = (IdentifierNameSyntax)memberInitializer;
+                        Debug.Assert(objectInitializerMemberBinder != null);
 
-                    // Bind member initializer assignment expression
-                    return BindAssignment(initializer, boundLeft, boundRight, isRef, diagnostics);
-                }
+                        var boundNode = objectInitializerMemberBinder.BindObjectInitializerMemberMissingAssignment(identifierName, implicitReceiver, diagnostics);
+                        // Do not bind to a non-member, like a BoundBadExpression
+                        if (boundNode.Kind is BoundKind.ObjectInitializerMember)
```

Comment:
>if (boundNode.Kind is BoundKind.ObjectInitializerMember) [](http://example.com/codeflow?start=24&length=56)

The logic here with all the rebinding and ```includeError``` flag feels unnecessary complicated. Also, it appears we are now creating a new tree shape whare an implicit receiver appears in a place where we didn't expect it to appear before (not under an assignment). Likely this is the reason for some of the adjustments made in `ControlFlowGraphBuilder.cs`. 

For the purpose of error recovery, I think it is reasonable to assume that a standalone identifier at this position is just an unfinished assignment and bind it as such. Meaning:
- use the ```boundNode``` regardless of what it is at this point
- create "WasCompilerGenerated" ```BoundAssignmentOperator```  with it as the target, "WasCompilerGenerated" ```BoundBadExpression``` with ```UnknownResultType```  as value
- return that assignment node

Reply:
I tried this approach and I can't figure out what syntax node to provide into the bound bad expression; either I use the identifier name twice, throwing when running the tests because the same syntax is used in multiple bound expressions, or I have to generate a syntax node that contains nothing and doesn't actually exist in the parsed tree.

Reply:
> I tried this approach and I can't figure out what syntax node to provide into the bound bad expression; either I use the identifier name twice, throwing when running the tests because the same syntax is used in multiple bound expressions, or ...

Did you accurately follow the suggestion? Did you mark the synthesized nodes as "WasCompilerGenerated"?

Reply:
I confused `hasError` with `WasCompilerGenerated`, it works fine now and I've applied this approach. Thanks for the guidance.

Reply:
This `if` is still present after commit 15. To clarify, the suggestion was to get rid of it and always return ```BoundAssignmentOperator```.

Reply:
I tried that and binding to the symbols failed, so this specific binding technique is necessary it seems

Reply:
> I tried that and binding to the symbols failed, so this specific binding technique is necessary it seems

Where did it fail, why did it fail? I assume it would work if the assignment was present in the syntax. In general, if a suggestion doesn't quite work, it would be good to explicitly communicate the fact.

Reply:
I tried again and removed the `if`, tests pass fine. This is done.